PR: https://github.com/odoo/odoo/pull/121193

From: bce4aededc37bac40c89aae44a157f20a873528f
From: Valeriya(vchu)
Date: 2023-06-12 15:40:58

Structural Changes: 2
Total Changes: 32

[IMP] (pos_,sale_)loyalty: improve validity period

currently it is only possible to define the end_date that
does not allow to prepare the program in advance.
Added the start_date property to the program model.
task-3305712

closes odoo/odoo#121193

Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_loyalty/models/pos_config.py
+++ b/addons/pos_loyalty/models/pos_config.py
@@ -87,22 +87,30 @@ class PosConfig(models.Model):
                 },
             }
         check_date = fields.Date.from_string(creation_date[:11])
-        if (coupon.expiration_date and coupon.expiration_date < check_date) or\
-            (coupon.program_id.date_to and coupon.program_id.date_to < fields.Date.context_today(self)) or\
-            (coupon.program_id.limit_usage and coupon.program_id.total_order_count >= coupon.program_id.max_usage):
-            return {
-                'successful': False,
-                'payload': {
-                    'error_message': _('This coupon is expired (%s).', code),
-                },
-            }
-        if not coupon.program_id.reward_ids or not any(reward.required_points <= coupon.points for reward in coupon.program_id.reward_ids):
+        today_date = fields.Date.context_today(self)
+        error_message = False
+        if (
+            (coupon.expiration_date and coupon.expiration_date < check_date)
+            or (coupon.program_id.date_to and coupon.program_id.date_to < today_date)
+            or (coupon.program_id.limit_usage and coupon.program_id.total_order_count >= coupon.program_id.max_usage)
+        ):
+            error_message = _("This coupon is expired (%s).", code)
+        elif coupon.program_id.date_from and coupon.program_id.date_from > today_date:
+            error_message = _("This coupon is not yet valid (%s).", code)
+        elif (
+            not coupon.program_id.reward_ids or
+            not any(r.required_points <= coupon.points for r in coupon.program_id.reward_ids)
+        ):
+            error_message = _("No reward can be claimed with this coupon.")
+
+        if error_message:
             return {
                 'successful': False,
                 'payload': {
-                    'error_message': _('No reward can be claimed with this coupon.'),
+                    'error_message': error_message,
                 },
             }
+
         return {
             'successful': True,
             'payload': {

--- a/addons/pos_loyalty/models/pos_session.py
+++ b/addons/pos_loyalty/models/pos_session.py
@@ -21,7 +21,7 @@ class PosSession(models.Model):
         return {
             'search_params': {
                 'domain': [('id', 'in', self.config_id._get_program_ids().ids)],
-                'fields': ['name', 'trigger', 'applies_on', 'program_type', 'date_to',
+                'fields': ['name', 'trigger', 'applies_on', 'program_type', 'date_from', 'date_to',
                     'limit_usage', 'max_usage', 'is_nominative', 'portal_visible', 'portal_point_name', 'trigger_product_ids'],
             },
         }
